cnudie/upload.py

```diff
@@ -0,0 +1,135 @@
+import enum
+import hashlib
+
+import ccc.oci
+import gci.componentmodel as cm
+import gci.oci
+import json
+import oci.model as om
+import dataclasses
+
+
+class UploadMode(enum.Enum):
+    SKIP = 'skip'
+    FAIL = 'fail'
+    OVERWRITE = 'overwrite'
+
+
+def _normalise_component_name(component_name: str) -> str:
+    return component_name.lower()  # oci-spec allows only lowercase
+
+
+def _target_oci_ref(
+    component: cm.Component,
+    component_ref: cm.ComponentReference=None,
+    component_version: str=None,
+):
+    if not component_ref:
+        component_ref = component
+        component_name = component_ref.name
+    else:
+        component_name = component_ref.componentName
+
+    component_name = _normalise_component_name(component_name)
+    component_version = component_ref.version
+
+    last_ctx_repo = component.current_repository_ctx()
+
+    return last_ctx_repo.component_version_oci_ref(
+        name=component_name,
+        version=component_version,
+    )
+
+
+def _sha_digest(fileobj) -> str:
```

Comment:
this function is new, is it not? I understand the idea of deduplicating the existing code, however, alas, I do have some concerns.

firstly, considering the name, this function calculates and returns a sha256--digest from passed-in object in hexadecimal form w/ `sha256:`-prefix. those things should be reflected in the name, e.g. `_sha256_hexdigest`

secondly, default chunksize should ideally be a parameter (it is okay to define a default-value, of course)

thirdly, not all filelike-objects are in fact seekable, and any modification to fpointer should ideally be mentioned in a docstr so callers do not have to read through the code (albeit it is short)

lastly, this function is at one place called w/ a byte-array (<class 'bytes'> as python will print its type); in this case, this will even break w/o special-case-handling depending on input-type.

from my gut-feeling, this is  little code and there is few deduplication (and too much complexity if this function would be "done right") to justify this factoring-out as a good idea.

we might factor-out the `sha256:`-prefixing. but then again: this is a contract brought upon us from Docker/OCI, and I strongly doubt it to ever change, so I do not see redundant hardcoding as an issue in this case.

Reply:
Agreed - removed